### PR TITLE
add CacheClearer

### PR DESCRIPTION
**Pull request by gam6itko**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| Doc updated   | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #834 
| License       | MIT

Works only with https://github.com/schmittjoh/metadata/pull/97

### CODE REVIEW

File: Cache/CacheClearer.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+{
+    /**
+     * @var ClearableCacheInterface
+     */
+    private $cache;
+
+    /**
+     * @param ClearableCacheInterface $cache
+     */
+    public function __construct($cache)
```
Comment (goetas) [inline Cache/CacheClearer.php:19]: Can you use here typehints and requre `CacheInterface`?

File: Cache/CacheClearer.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        $this->cache = $cache;
+    }
+
+    public function clear(string $cacheDirectory): void
+    {
+        if (method_exists($this->cache, 'clear')) { // $this->cache instanceof ClearableCacheInterface
```
Comment (goetas) [inline Cache/CacheClearer.php:26]: why not check `$this->cache instanceof ClearableCacheInterface` ?

**Review by goetas — CHANGES_REQUESTED** [review]
i've released your changes in the metadata package as 2.5.0 (see https://github.com/schmittjoh/metadata/releases/tag/2.5.0)

can you add it as requirement in the composer.json?

File: Cache/CacheClearer.php (hunk 1 above)
Comment (gam6itko) [reply Cache/CacheClearer.php:19]: I'll do it

**Comment by gam6itko** [comment]
@goetas I made the changes you asked for

**Comment by goetas** [comment]
thank you!

**Comment by gam6itko** [comment]
@goetas Is it acceptable that it failed the test with --prefer-lowet?

**Comment by goetas** [comment]
Hmm, I did not notice that... Ideally it should be fixed... 😔

**Comment by gam6itko** [comment]
@goetas problem in CacheClearerTest. I can rewrite it according to `--prefer-lowest`

File: Resources/config/services.xml [hunk 3]
```diff
@@ -233,6 +233,11 @@
             <tag name="jms_serializer.deserialization_visitor" format="xml" />
         </service>
 
+        <service id="jms_serializer.cache.cache_clearer" class="JMS\SerializerBundle\Cache\CacheClearer" public="false">
+            <argument type="service" id="jms_serializer.metadata.cache"/>
```
Comment (marmichalski) [inline Resources/config/services.xml:237]: I think this is missing `on-invalid="null` entry, as disabling cache (with `cache: none`) is broken now.

The constructor should be updated to accept null then. :)

Comment (gam6itko) [reply Resources/config/services.xml:237]: I'll fix it ASAP

Comment (marmichalski) [reply Resources/config/services.xml:237]: I have already opened a PR #850., would be nice if you could review it. 👍

**Pull request merged**